thetis/timeintegrator.py

```diff
@@ -433,6 +433,127 @@ def advance(self, t, update_forcings=None):
             self.fields_old[k].assign(self.fields[k])
 
 
+class CoupledTracerPicard(TimeIntegrator):
+    r"""
+    Picard iteration for coupled tracer equations.
+
+    Given an :class:`OrderedDict` of equations and an iteration count,
+    timestep each equation in sequence at each iteration. Timestepping
+    of each component is achieved using Crank-Nicolson with implictness
+    :class:`\theta`.
+
+    This :class:`TimeIntegrator` is designed for use with systems of
+    tracer equations. If the equations are all of advection-diffusion
+    type, then the order does not matter and only one iteration is
+    required. If reaction terms are included which provide couplings
+    between the equations, then the order is important and it may be
+    beneficial to take multiple iterations per timestep.
+    """
+    cfl_coeff = CFL_UNCONDITIONALLY_STABLE  # We have Crank-Nicolson in each component
+
+    def __init__(self, equations, solutions, fields, dt,
+                 bnd_conditions=None, solver_parameters=None,
+                 theta=0.5, semi_implicit=False, iterations=1):
+        """
+        :arg equations: Dictionary of :class:`Equation` objects.
+        :arg solutions: Dictionary of :class:`Function` objects.
+        :arg fields: Dictionary of fields that are passed to each equation
+        :arg float dt: time step in seconds
+        :kwarg dict bnd_conditions: Dictionary of boundary conditions passed to each equation
+        :kwarg dict solver_parameters: PETSc solver options
+        :kwarg theta: Implicitness parameter, which can be given as either a single value or
+            a dictionary corresponding to the different tracer fields
+        :kwarg bool semi_implicit: If True use a linearized semi-implicit scheme
+        :kwarg int iterations: Number of Picard iterations
+        """
+        super(CoupledTracerPicard, self).__init__(equations, solutions, fields, dt, solver_parameters)
+        self.equations = equations
+        self.solutions = solutions
+        delattr(self, 'equation')
+        delattr(self, 'solution')
```

Comment:
Hmm, what you have done here suggest that this is not a TimeIntegrator, but rather a CoupledTimeintegrator (as it handles multiple equations). Can this not be designed as one? That is, is there any difference between the present implementation and bundling up a a few CrNi Timeintegrators and stepping them in the right order? You've also re-implemented the Cr-Ni forms which is redundant (unless I'm missing something). So I'd move this to `coupled_timeintegrator_2d.py`. Need to think about the names I guess. This class only handles tracer equations, whereas `CoupledTimeIntegrator2D` and offspring also include swe (and sediments).

Reply:
I think the only difference is making sure that the solves are interleaved during `advance`. i.e.
```
for it in range(self.iterations):
    for solver in self.solvers:
        solver.solve()
```
rather than
```
for solver in self.solvers:
    for it in range(self.iterations):
        solver.solve()
```